content/blog/2023-12-01-task-lifecycle.md

```diff
@@ -0,0 +1,572 @@
+  kernel_execve()
+    bprm_execve()
+      exec_binprm()
+        search_binary_handler()
+          list_for_each_entry(fmt, &formats, lh)
+            retval = fmt->load_binary(bprm);
+```
+
+The outer few calls mainly handle checking prerequisites and bookkeeping. The
+`exec_binrpm()` call also handles shebang redirection, allowing up to 5 levels
+of interpretor. At each level it invokes `search_binary_handler()`, which
+attempts to find a handler for the program file's format. Contrary to the name,
+the searcher will also immediately try to load the file if it finds an
+appropriate handler. It's this call to `load_binary` (dispatched to whatever
+handler was found) that sets up our userspace execution context, including the
+syscall return state.
+
+All that's left to do here is return 0 all the way up the chain, which you'll
+see results in the init task returning to the shim that performs the syscall
+return sequence to userspace. The init task is now fully userspace.
+
+
+## Creating other tasks
+
+It feels like we've spent a lot of time discussing the init task. What about all
+the other tasks?
+
+It turns out that the creation of the init task is very similar to any other
+task. All tasks are clones of the task that created them (except the statically
+defined `init_task`). Note 'clone' is being used in a loose sense here: it's not
+an exact image of the parent. There's a configuration parameter that determines
+which components are shared, and which are made into independent copies. The
+implementation may also just decide to change some things that don't make sense
+to duplicate, such as the task ID to distinguish it from the parent.
+
+As we saw earlier, kthreads are created indirectly through a global list and
+kthreadd daemon task that does the actual cloning. This has two benefits:
+allowing asynchronous task creation from atomic contexts, and ensuring all
+kthreads inherit a 'clean' task context, instead of whatever was active at the
+time.
+
+Userspace task creation, beyond the init task, is driven by the userspace
+process invoking the `fork()` and `clone()` family of syscalls. Both of these
+are light wrappers over the `kernel_clone()` function, which we used earlier for
+the creation of the init task and kthreadd.
+
+When a task runs a syscall in the `exec()` family, it doesn't create a new task.
+It instead hits the same code path as when we tried to run the userspace init
+program, where it loads in the context as defined by the program file into the
+current task and returns from the syscall (legitimately this time).
+
+
+## Context switching
+
+The last piece of the puzzle (as far as this article will look at!) is how tasks
+are switched in and out, and some of the rules around when it can and can't
+happen. Once the init and kthreadd tasks are created, we call
+`cpu_startup_entry(CPUHP_ONLINE)`. Any coprocessors have also been released to
+call this by now too. These are considered 'idle tasks', which serve to run when
```

Comment:
"These are considered" I assume you mean that calling cpu_startup_entry() creates an idle task for that CPU?